File: Providers/Localytics/AnalyticsKitLocalyticsProvider.m
//
//  AnalyticsKitLocalyticsProvider.m
//  AnalyticsKit
//
//  Created by Todd Huss on 10/17/11.
//  Copyright (c) 2011 Two Bit Labs. All rights reserved.
//

#import "AnalyticsKitLocalyticsProvider.h"
#import "Localytics.h"

@implementation AnalyticsKitLocalyticsProvider

-(id<AnalyticsKitProvider>)initWithAPIKey:(NSString *)localyticsKey {
    self = [super init];
    if (self) {
        [Localytics integrate:localyticsKey];
        [Localytics openSession];
    }
    return self;
}

-(void)applicationWillEnterForeground {
    [Localytics openSession];
    [Localytics upload];
}

-(void)applicationDidEnterBackground {
    [Localytics closeSession];
    [Localytics upload];
}

-(void)applicationWillTerminate { 
    [Localytics closeSession];
    [Localytics upload];
}

-(void)uncaughtException:(NSException *)exception {
    [Localytics tagEvent:@"Uncaught Exceptions" attributes:
     [NSDictionary dictionaryWithObjectsAndKeys:
          [exception name], @"ename",
          [exception reason], @"reason",
          [exception userInfo], @"userInfo",
      nil]];

}

-(void)logScreen:(NSString *)screenName {    
    [Localytics tagScreen:screenName];
}

-(void)logEvent:(NSString *)event {
    [Localytics tagEvent:event];
}

-(void)logEvent:(NSString *)event withProperties:(NSDictionary *)dict {
    [Localytics tagEvent:event attributes:dict];
}

-(void)logEvent:(NSString *)event withProperty:(NSString *)key andValue:(NSString *)value {
    [Localytics tagEvent:event attributes:[NSDictionary dictionaryWithObject:value forKey:key]];
}

-(void)logEvent:(NSString *)eventName timed:(BOOL)timed {
    [self logEvent:eventName];
}

-(void)logEvent:(NSString *)eventName withProperties:(NSDictionary *)dict timed:(BOOL)timed {
    [self logEvent:eventName withProperties:dict];
}

-(void)endTimedEvent:(NSString *)eventName withProperties:(NSDictionary *)dict {}

-(void)logError:(NSString *)name message:(NSString *)message exception:(NSException *)exception {
    [Localytics tagEvent:@"Exceptions" attributes:
     [NSDictionary dictionaryWithObjectsAndKeys:
        name, @"name",
        message, @"message",        
        [exception name], @"ename",
        [exception reason], @"reason",
        [exception userInfo], @"userInfo",
      nil]];
}

-(void)logError:(NSString *)name message:(NSString *)message error:(NSError *)error {
    [Localytics tagEvent:@"Errors" attributes:
     [NSDictionary dictionaryWithObjectsAndKeys:
          name, @"name",
          message, @"message",        
          [error localizedDescription], @"description",
          [NSString stringWithFormat:@"%ld", (long)[error code]], @"code",
          [error domain], @"domain",
          [[error userInfo] description], @"userInfo",
      nil]];
}

@end

File: Providers/Localytics/AnalyticsKitLocalyticsProvider.swift
import Foundation
import Localytics

public class AnalyticsKitLocalyticsProvider: NSObject, AnalyticsKitProvider {
    @objc(initWithAPIKey:)
    public init(apiKey localyticsKey: String) {
        Localytics.integrate(localyticsKey)
        Localytics.openSession()
    }

    public func applicationWillEnterForeground() {
        Localytics.openSession()
        Localytics.upload()
    }

    public func applicationDidEnterBackground() {
        Localytics.closeSession()
        Localytics.upload()
    }

    public func applicationWillTerminate() {
        Localytics.closeSession()
        Localytics.upload()
    }

    public func uncaughtException(_ exception: NSException) {
        Localytics.tagEvent("Uncaught Exceptions", attributes: [
            "ename": exception.name.rawValue,
            "reason": exception.reason ?? "nil",
        ])
    }

    public func logScreen(_ screenName: String) {
        Localytics.tagScreen(screenName)
    }

    public func logScreen(_ screenName: String, withProperties properties: [String : Any]) {
        logScreen(screenName)
    }

    public func endTimedEvent(_ event: String, withProperties properties: [String: Any]) {}

    public func logEvent(_ event: String) {
        Localytics.tagEvent(event)
    }
    
    public func logEvent(_ event: String, withProperties properties: [String : Any]) {
        Localytics.tagEvent(event, attributes: properties as? [String : String])
    }

    public func logEvent(_ event: String, withProperty key: String, andValue value: String) {
        logEvent(event, withProperties: [key: value])
    }

    public func logEvent(_ event: String, timed: Bool) {
        logEvent(event)
    }

    public func logEvent(_ event: String, withProperties properties: [String : Any], timed: Bool) {
        logEvent(event, withProperties: properties)
    }

    public func logError(_ name: String, message: String?, properties: [String: Any]?, exception: NSException?) {
        var attributes: [String: String] = [
            "name": name,
            "message": message ?? "nil",
            "ename": exception?.name.rawValue ?? "nil",
            "reason": exception?.reason ?? "nil",
        ]
        if let properties = properties {
            let stringsDict = Dictionary(uniqueKeysWithValues: properties.map { ($0, "\($1)") } )
            attributes.merge(stringsDict) { (current, _) in current }
        }
        Localytics.tagEvent("Exceptions", attributes: attributes)
    }

    public func logError(_ name: String, message: String?, properties: [String: Any]?, error: Error?) {
        var attributes: [String: String] = [
            "name": name,
            "message": message ?? "nil",
            "description": error?.localizedDescription ?? "nil",
        ]
        if let properties = properties {
            let stringsDict = Dictionary(uniqueKeysWithValues: properties.map { ($0, "\($1)") } )
            attributes.merge(stringsDict) { (current, _) in current }
        }
        Localytics.tagEvent("Exceptions", attributes: attributes)
    }
}
